src/swoop/api/routers/payloads.py

```diff
@@ -22,16 +43,64 @@
     response_model=PayloadList,
     responses={"404": {"model": APIException}},
 )
-def list_payloads(
-    limit: int = Query(ge=1, default=DEFAULT_JOB_LIMIT),
+async def list_payloads(
+    request: Request,
+    limit: int = Query(ge=1, default=DEFAULT_PAYLOAD_LIMIT),
     process_id: list[str] | None = Query(default=None),
     collection_id: list[str] | None = Query(default=None),
     item_id: list[str] | None = Query(default=None),
 ) -> PayloadList | APIException:
     """
     retrieve the list of payloads.
     """
-    pass
+    proc_clause = V("c.workflow_name") == funcs.any(process_id)
+    coll_clause = V("i.collection") == funcs.any(collection_id)
+    item_clause = V("i.item_id") == funcs.any(item_id)
+
+    async with request.app.state.readpool.acquire() as conn:
+        q, p = render(
+            """
+                WITH item_cache AS (
+                    SELECT c.payload_uuid
+                    FROM swoop.item_payload p
+                    INNER JOIN swoop.payload_cache c
+                    ON c.payload_uuid = p.payload_uuid
+                    INNER JOIN swoop.input_item i
+                    ON p.item_uuid = i.item_uuid
+                    WHERE (:processes::text[] IS NULL OR :proc_where)
+                    AND (:collections::text[] IS NULL OR :coll_where)
+                    AND (:items::text[] IS NULL OR :item_where)
+                    LIMIT :limit::integer
+                )
+                SELECT
+                    DISTINCT(payload_uuid)
+                FROM
+                    item_cache;
```

Comment:
I don't think the `limit` is going to work the way you expect. For example, if your `limit` is 10 and you have 10 items for a single payload, you will only get one payload ID returned once the `distinct` is applied, right?

I also had a bit different idea of how the CTE would work; seeing the query you've ended up with here, I am thinking the guidance to consider a CTE was a bit of a misdirection. I don't think the CTE does anything here except complicate matters. Why not remove it entirely?

Here's a suggestion (with some stylistic changes too):

```suggestion
                SELECT
                  DISTINCT(c.payload_uuid)
                FROM swoop.item_payload AS p
                INNER JOIN swoop.payload_cache AS c USING (payload_uuid)    
                INNER JOIN swoop.input_item AS i USING (item_uuid)
                WHERE
                  (:processes::text[] IS NULL OR :proc_where)
                  AND (:collections::text[] IS NULL OR :coll_where)
                  AND (:items::text[] IS NULL OR :item_where)
                LIMIT :limit::integer
```

I do think that all the joins are a performance concern when filter _is not_ applied, which is what I was thinking about with the CTE idea initially: you can create a CTE, let's call it `payload_ids` for example purposes, to return distinct payload IDs based on any provided item ID/collection filters _only if those filters are specified_. Then you can use a `payload_id in (select * from payload_ids)` where clause that is added only when you have the `payload_ids` CTE.

But that idea is perhaps premature optimization, what you have here (with something like the tweaks I suggest above) is sufficient for now.

Reply:
It's probably worth adding a test case triggering the `limit` issue I mentioned before trying to fix it. You don't need to use a `limit` of 10: a `limit` of 2 with three items, two for one payload, should be sufficient to produce the issue.